% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/MADloy.R
\name{madloy}
\alias{madloy}
\title{Check the mean LRR values in msY region to detect Loss of Y events in a folder or files.}
\usage{
madloy(
  files,
  target.region,
  ref.region = "Autosomes",
  qc.sds = 0.28,
  rsCol = 1,
  ChrCol = 2,
  PosCol = 3,
  LRRCol = 4,
  trim = 0.05,
  offset,
  mc.cores,
  quiet = FALSE,
  hg = "GRCh38",
  ...
)
}
\arguments{
\item{files}{A single file path (APT platform and MAD platform), a vector of 
file paths (MAD platform) or a MAD rawData folder path containing files 
ready to be processed with MAD (MAD platform).}

\item{target.region}{The chromosome or region to be compared with the other 
regions. By default is the region chrY:2694521-59034049 (hg19/GRCh37) but it can be 
changed.}

\item{ref.region}{If declared, the chromosome or region to be compared with the Y 
region in UCSC style (i.e. "chr21" or "chr21:1000-10000").}

\item{qc.sds}{Theshold to perform quality control of samples using the reference
region/chromosome. The default value is 0.28. It implies that samples having 
LRR standard deviation larger that this value in the reference chromosome will be
removed from the analyses. This value is taken from 
www.illumina.com/content/dam/illumina-marketing/documents/products/appnotes/appnote_cnv_loh.pdf)
If NULL the cutoff for considering a good quality sample is estimated as 
2 times the standard deviation of all samples.}

\item{rsCol}{The position of the column with the SNP identifier.}

\item{ChrCol}{The position of the column with the Chromosome field.}

\item{PosCol}{The position of the column with the Position field.}

\item{LRRCol}{The position of the column with the LRR identifier.}

\item{trim}{trim the fraction (0 to 0.5) of probes to be trimmed when summaryzing LRR. By default is 0.05.}

\item{offset}{median value of summarized LRR at target region}

\item{mc.cores}{The number of cores used to perform the function. By default 
is set to 1.}

\item{quiet}{Should the function not inform about the status of the process. 
By default is FALSE.}

\item{hg}{Human genome build version. It can be 'hg18', 'hg19' or 'GRCh38'. Set by default to 'GRCh38'.}

\item{...}{Other parameters.}
}
\value{
A MADloy object that contains the LRR means for all the files 
  analyzed.
}
\description{
madloy check the median log R ratio(LRR) of all the MAD files specified or in
a path to detect Loss of Y events. The LRR is computed by default for the 
autosomes and msY chrY:2694521-59034049 (hg19/GRCh37).
}
\examples{
\dontrun{
madloy(filepath, mc.cores=2)}
}
\seealso{
\code{\link{getLOY}} to process results from \code{MADloy}
}
